server/src/main/java/org/elasticsearch/common/logging/DeprecationLogger.java

```diff
@@ -261,12 +261,19 @@ public Void run() {
     }
 
     public String getXOpaqueId(Set<ThreadContext> threadContexts) {
-        return threadContexts.stream()
-                             .filter(t -> t.isClosed() == false)
-                             .filter(t -> t.getHeader(Task.X_OPAQUE_ID) != null)
-                             .findFirst()
-                             .map(t -> t.getHeader(Task.X_OPAQUE_ID))
-                             .orElse("");
+        for (ThreadContext threadContext : threadContexts) {
+            try {
+                if (threadContext.isClosed() == false) {
+                    String header = threadContext.getHeader(Task.X_OPAQUE_ID);
+                    if (header != null) {
+                        return header;
+                    }
+                }
+            } catch (IllegalStateException e) {
+                // ignore exception as this is due to a race condition  between isClosed and getHeader. Only in test env.
```

Comment:
Can we rethrow this if we aren't in a test environment? Could we fix the race condition? Blanket hiding of exceptions like this is dangerous because other exception sources can sneak in over time and we would be masking them unknowingly.

Reply:
good point. Fixing this in `ThreadContext` would require synchronisation and would impose performance penalty. 
The minimum would be a  ReadWrite lock  between close(write) and get(read).

I wonder if we should fix this in `CloseableThreadLocal` where the NullPointerException is thrown from
the `close()` method nullifies the `threadLocal` variable `  t = null;`
```
 @Override
  public void close() {
 ...
    hardRefs = null;
  ...
    if (t != null) {
      t.remove();
    }
    t = null;
  }
```
but then first thing the `get()` method does is to access it without checking if it is null 
```
public T get() {
    WeakReference<T> weakRef = t.get();
...
```
if we modify `t` and `hardRefs` to be a volatile variables and  assign with "null objects" (new ThreadLocal and new empty map) or
we still nullify but we also guard `t` and `hardRefs` in `get()` and `set()` against nulls it should be fixed.

I wonder if this was done on purpose to keep the design of the class simple or I am missing something. The `ClosableThreadLocal` warns about this in java doc.
```
You should not call {@link #close} until all
 *  threads are done using the instance.
```

I wonder if it is not too expensive to introduce synchronisation (either a lock or volatile) just because of testability. Again, we cannot even assert about the values of `XOpaqueId` when run in parallel. Only when there is only 1 Node per jvm (=> 1 `ThreadContext per jvm) we can reliably assert about these values

Reply:
For now I would be tempted to throw exception on Prod but I am also happy to create a new issue to possibly fix this 
WDYT?

Reply:
Yes I would prefer to not mask the exception in non-test environments since it is not expected there.

Reply:
@s1monw what are your views on fixing this in lucene?

Reply:
I don't think we should change any of this code. IMHO the problem is that we pass in a threadlocal / threadcontext that is closed concurrently? How is this possible? This entire usage is desiged to be not accessed after it's closed, we should stick to it. Can you explain why this fails and maybe can we see an exception? I think the test must be fixed instead. This will just hide exceptions and cause trouble.

Reply:
@s1monw I think   the concurrent closure can only happen in **Test** environment. See here https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/common/logging/DeprecationLogger.java#L61 
where we have a `Set` of `ThreadContext` and some of them can be closed. In **prod** I expect only one `ThreadContext`, but I guess `close()` can still happen on a different thread. 

The stacktrace can be found here https://github.com/elastic/elasticsearch/issues/45443

 the reason it ends up with this race condition is because it iterates over set of `ThreadContext` then it checks if it is closed, and then attempts to retrieve a header value. However closed and get are not synchronized together, so it might happen that a thread1 calls `isClosed -> false` then, thread2 calls `close` and thread1 comes back and calls `getHeader` which fails as in meantime `ThreadContext` was closed by thread2

In my usecase, it would be fine if I modify `ThreadContext` to return some default or null when `get` is called but it was already closed. But not sure of other scenarios. 

I suppose this problem has been seen earlier as per comment here https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/common/logging/DeprecationLogger.java#L241
```
  } catch (final IllegalStateException e) {
                    // ignored; it should be removed shortly
                }
```

Reply:
alright, I understand now. I think the design of this is flawed but it's very hard to fix. Ideally we should refactor `DeprecationLogger` to only use 1 `ThreadContext` use `SetOnce` I think that's wise. Now that might be tricky if we allow to start multiple nodes in the same environment. 

What I think we should do instead is to use a read-write lock and acquire the read-lock every-time we access that `THREAD_CONTEXT` set. One we access it for write we get exclusive access and for reading we can be concurrent. The write access will only be acquired once in a production env which is good. All we need to make sure now is that we close the `ThreadContext` after we removed it. Does this make sense? Then you can also remove https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/common/logging/DeprecationLogger.java#L241